### PR TITLE
Add timeout to saltnado's event listener

### LINKED ISSUE #21707: [saltnado rest] local requests hang indefinitely on glob / grain queries
I'm making a rubygem handler for Salt REST API, and for some reason whenever I use a glob or grain query it'll stall indefinitely.

I'm using Typhoeus to make HTTP requests.

The request code looks like this:
```ruby
client.lowstate.local(fun: 'test.ping', tgt: '*')

# Request
#<Typhoeus::Request:0x007ff92b10b4a0
 @base_url="localhost:8000/",
 @options=
  {:method=>:post,
   :body=>{:client=>"local", :tgt=>"*", :fun=>"test.ping", :expr_form=>"glob"},
   :headers=>
    {"User-Agent"=>"Typhoeus - https://github.com/typhoeus/typhoeus",
     :Accept=>"application/json",
     :ContentType=>"application/json",
     "X-Auth-Token"=>"token"},
   :maxredirs=>50},
 @original_options=
  {:method=>:post,
   :body=>{:client=>"local", :tgt=>"*", :fun=>"test.ping", :expr_form=>"glob"},
   :headers=>
    {:Accept=>"application/json",
     :ContentType=>"application/json",
     "X-Auth-Token"=>"token"}}>

# body: "client=local&tgt=*&fun=test.ping&expr_form=glob"
# url:  "localhost:8000/"
```

I can do the same thing on ``local_async`` and it'll work but I have no way of knowing when the job is actually done.

Any ideas?

### PR DESCRIPTION
**Pull request by jacksontj**
Fixes #21707

The issue is basically that the master has seen N minions-- and not all N are currently connected. The publish job returns to the salt-api that N minions got the job-- and it dutifilly waits for all the returns. With this we use the same ping timeout that the CLI does while waiting for returns. So once all minions are no longer running the job we will return.

### CODE REVIEW

**Comment by jacksontj** [comment]
@rallytime or @jfindlay should I open this PR against 2015.2 as well? Or will this get merged forward auto-magically?

**Comment by jfindlay** [comment]
Merge forwards happen whenever Colton gets a few free minutes; :-) usually once every 2-4 days.

**Comment by jacksontj** [comment]
@jfindlay To make sure it isn't missed in 2015.2 I just opened #21756 :) I figure that'll be easier

**Comment by jfindlay** [comment]
Sure, thanks @jacksontj, great work as usual.

**Comment by whiteinge** [comment]
@jacksontj ooc, could you lean on the existing timeout semantics in LocalClient() instead of implementing it here? We moved the `LocalClient().cmd()` method to use the same timeout behavior as the CLI for the 2014.7 release.

**Comment by jacksontj** [comment]
While I'm in here we'll go ahead and backport that big performance fix ;)

**Comment by jacksontj** [comment]
@whiteinge Depends on what you mean by that-- this is basically making it have the same timeout semantics as localclient and the CLI (before it was a mess). The only real change here was that we assumed that the Publisher's list was exhaustive and accurate-- so we would never timeout waiting for a ping return of a minion-- which in the case of a dead one means you'll wait until someone turns it back on.

If you mean "can we use the same code" that gets more complicated-- since saltnado is all ioloop/async stuff it doesn't quite work the same-- but the logic here is equivalent-- just lots less code ;) I have a branch on my github where I'm moving most of the internals over to tornado stuff-- so we'll be able to consolidate the timeout code once that's done.

**Comment by whiteinge** [comment]
Roger that (sorry for the brevity in my last comment; wild day today). You're very right that it was a mess -- `cmd()` and Salt's CLI had badly diverging behavior. In 2014.7 I moved `LocalClient().cmd()` to use `get_cli_event_returns` so it should be identical to the CLI now and hopefully not a mess. (Or at least a centrally-managed mess.)

Good point that Tornado is coming at this from an async perspective. I don't know how well the existing `get_iter_returns` (which `get_cli_event_returns` wraps) works with that kind of system (if at all). I know some work was done to get it Gevent-friendly a while back but that's a very different async strategy.

Nevermind me. :walking:

**Comment by jacksontj** [comment]
Lol, no worries ;) it is a bit of a mess. Now (as you mentioned) its a central mess, but a mess nonetheless. This implementation has the same outcome but just has less code-- plan is to replace the ugly mess with this code (or the equivalent), but it's part of a larger re-write ;) We'll get there :)


It might be nice to make a test case for testing command returns-- that way we can test that cmd(), api, and cli have the same outcome-- since we can't really gaurantee that they'll use the same code forever-- specifically it is probably worth making a minion in the integration suite which isn't on (since thats the case we are testing).

**Comment by whiteinge** [comment]
+1

File: salt/runner.py [hunk 1]
```diff
@@ -53,6 +53,7 @@ def cmd(self, fun, arg, pub_data=None, kwarg=None):
 
         .. code-block:: python
 
+<<<<<<< HEAD
```
Comment (cachedout) [inline salt/runner.py]: Could you pull this out please @jacksontj ?

**Comment by jacksontj** [comment]
@cachedout Done

**Comment by cachedout** [comment]
Thank you @jacksontj

**Comment by baweaver** [comment]
Could we look at potentially adding something to the README section on this one? I assume clients would just send something to the equivalent of ``-d timeout=n``?

**Comment by jacksontj** [comment]
@whiteinge Now you have me all worried about the timeouts not matching up. I'll take a third look at this later today ;)

File: salt/netapi/rest_tornado/saltnado.py [hunk 2]
```diff
@@ -294,15 +290,30 @@ def get_event(self,
...
-                response = future.result()
-                tornado.ioloop.IOLoop.current().add_callback(callback, response)
+                tornado.ioloop.IOLoop.current().add_callback(callback, future)
             future.add_done_callback(handle_future)
         # add this tag and future to the callbacks
         self.tag_map[tag].append(future)
         self.request_map[request].append((tag, future))
 
+        if timeout:
+            timeout_future = tornado.ioloop.IOLoop.current().call_later(timeout, self._timeout_future, tag, future)
```
Comment (isbm) [inline salt/netapi/rest_tornado/saltnado.py]: Sorry for the nitpick, but this probably wants to be wrapped a little. :)

File: salt/netapi/rest_tornado/saltnado.py [hunk 3]
```diff
@@ -896,7 +910,7 @@ def _disbatch_local(self, chunk):
                 ping_id = event['data']['id']
                 if ping_id not in chunk_ret and ping_id not in minions_remaining:
                     minions_remaining.append(ping_id)
-                ping_event = self.application.event_listener.get_event(self, tag=ping_tag)
+                ping_event = self.application.event_listener.get_event(self, tag=ping_tag, timeout=self.application.opts['gather_job_timeout'])
```
Comment (isbm) [inline salt/netapi/rest_tornado/saltnado.py]: This wants to be wrapped too.

File: salt/netapi/rest_tornado/saltnado.py [hunk 4]
```diff
@@ -881,7 +895,7 @@ def _disbatch_local(self, chunk):
         minions_remaining = pub_data['minions']
 
         ret_event = self.application.event_listener.get_event(self, tag=ret_tag)
-        ping_event = self.application.event_listener.get_event(self, tag=ping_tag)
+        ping_event = self.application.event_listener.get_event(self, tag=ping_tag, timeout=self.application.opts['gather_job_timeout'])
```
Comment (isbm) [inline salt/netapi/rest_tornado/saltnado.py]: And this line is too long.

**Comment by jacksontj** [comment]
@whiteinge Yup, we didn't handle the case where a minion started with the job, then lost it (minion died, job stopped, w/e). #21811 includes the fixes for that-- which I think ends up being a nicer interface-- since its a single future that you are returned :)

**Pull request merged**